importer/balanced/builder.go

```diff
@@ -1,24 +1,40 @@
-// Package balanced provides methods to build balanced DAGs.
-// In a balanced DAG, nodes are added to a single root
-// until the maximum number of links is reached (with leaves
-// being at depth 0). Then, a new root is created, and points to the
-// old root, and incorporates a new child, which proceeds to be
-// filled up (link) to more leaves. In all cases, the Data (chunks)
-// is stored only at the leaves, with the rest of nodes only
-// storing links to their children.
+// Package balanced provides methods to build balanced DAGs. In a
+// balanced DAG, nodes are added to a single root until the maximum
+// number of links is reached (with leaves being at depth 0). Then, a new
+// root is created to point to the old root (incorporating it as its
+// first child), which proceeds to be filled up (link) to more nodes. In
+// all cases, the data (chunk) is stored only at the leaves, with the
+// rest of (internal) nodes only storing links to their children and
+// sizes of the data stored under them.
 //
-// In a balanced DAG, nodes fill their link capacity before
-// creating new ones, thus depth only increases when the
-// current tree is completely full.
+// Internal nodes are always represented by UnixFS nodes (of type `File`)
+// encoded inside DAG nodes, in contrast, leaf nodes with data have
+// multiple possible representations: UnixFS nodes as above, raw nodes
+// with just the file data (no format) and Filestore nodes (supported by
+// raw nodes). In the case the entire file fits into just one node it
+// will be formatted as a (single) leaf node (without parent) with the
+// possible representations already mentioned, this is the only scenario
+// where the root can be of a type different that the UnixFS node.
 //
-// Balanced DAGs are generalistic DAGs in which all leaves
-// are at the same distance from the root.
+// The nodes are filled by recursion (in a DFS way) so the DAG is built
+// from the bottom up. Leaf nodes are created first with the file data
+// and its size that is bubbled up to the parent (internal) node, that in
+// turn will aggregate all the file sizes of its children and bubble them
+// up to its parent (to keep propagating the total file size up to the
+// root). In this way the DAG acts like a B-tree (a more appropriate data
+// structure may be in order for this analogy), with each internal node
+// using the file size of its children as an index to perform the search
+// (seek) operation.
```

Comment:
Perfect.

Reply:
Then yes, I better lose the DFS term, because that will clash with the explanation of the actual search, which is not of a DFS type, I'll just put an emphasis on the recursive aspect of the construction (that doesn't belong to the search).